pliers/extractors/microsoft.py

```diff
@@ -0,0 +1,182 @@
+        else:
+            attributes = ''
+
+        params = {
+            'returnFaceId': self.face_id,
+            'returnFaceLandmarks': self.landmarks,
+            'returnFaceAttributes': attributes
+        }
+        raw = self._query_api(stim, params)
+        return ExtractorResult(None, stim, self, raw=raw)
+
+    def _parse_response_json(self, json):
+        data_dict = {}
+        for k, v in json.items():
+            if k == 'faceRectangle' and not self.rectangle:
+                continue
+            if k == 'faceAttributes':
+                k = 'face'
+            if isinstance(v, dict):
+                subdata = self._parse_response_json(v)
+                for sk, sv in subdata.items():
+                    data_dict['%s_%s' % (k, sk)] = sv
+            elif isinstance(v, list):
+                # Hard coded to this extractor
+                for attr in v:
+                    if k == 'hairColor':
+                        key = attr['color']
+                    elif k == 'accessories':
+                        key = '%s_%s' % (k, attr['type'])
+                    else:
+                        continue
+                    data_dict[key] = attr['confidence']
+            else:
+                data_dict[k] = v
+        return data_dict
+
+    def _to_df(self, result):
+        face_results = []
+        for i, face in enumerate(result.raw):
+            face_data = self._parse_response_json(face)
+            face_results.append(face_data)
+
+        return pd.DataFrame(face_results)
+
+
+class MicrosoftAPIFaceEmotionExtractor(MicrosoftAPIFaceExtractor):
+
+    ''' Extracts facial emotions from images using the Microsoft API '''
+
+    def __init__(self, face_id=False, rectangle=False, landmarks=False,
+                 **kwargs):
+        super(MicrosoftAPIFaceEmotionExtractor, self).__init__(face_id,
+                                                               rectangle,
+                                                               landmarks,
+                                                               ['emotion'],
+                                                               **kwargs)
+
+
+class MicrosoftVisionAPIExtractor(MicrosoftVisionAPITransformer,
+                                  ImageExtractor):
+    ''' Base MicrosoftVisionAPIExtractor class.
+
+    Args:
+        features (list): one or more specified vision features as strings.
+            Supported vision features include Tags, Categories, ImageType,
+            Color, and Adult. Note that each attribute has additional
+            computational and time cost. By default extracts all visual
+            features from an image.
+    '''
+
+    api_method = 'analyze'
+    _log_attributes = ('api_version', 'features')
+
+    def __init__(self, features=None, **kwargs):
+        if hasattr(self, '_feature'):
+            self.features = [self._feature]
+        else:
+            self.features = features if features else ['Tags', 'Categories',
```

Comment:
Maybe add a check here to make sure any passed values are valid--unless the API ignores invalid ones, in which case it doesn't matter much. Though it's probably still a good idea to raise an exception on invalid values before any API calls are made.

Reply:
The Extractor will raise an exception when `.transform` is called (the API will return an error code). That ok or should we do it on `__init__`?

Reply:
If the error is comprehensible to the user, it's fine as is. If not, we should do it on init.

Reply:
The message is `Specified feature type is not valid` so I think that's fine